File: frontend/src/services/paymentService.js
import api from './api';

export const paymentService = {
  /**
   * Creates a payment order for an application order.
   * Endpoint: POST /api/payments/create/{orderId}
   */
  createPayment: async (orderId, paymentData = {}) => {
    const response = await api.post(`/payments/create/${orderId}`, paymentData);
    return response.data;
  },

  /**
   * Verifies Razorpay payment signature.
   * Endpoint: POST /api/payments/verify
   */
  verifyPayment: async (verifyData) => {
    const response = await api.post('/payments/verify', verifyData);
    return response.data;
  },

  /**
   * Retrieves payment details for an order.
   * Endpoint: GET /api/payments/order/{orderId}
   */
  getPaymentByOrder: async (orderId) => {
    const response = await api.get(`/payments/order/${orderId}`);
    return response.data;
  },

  /**
   * Requests a refund for a successfully paid order.
   * Endpoint: POST /api/payments/refund/{orderId}
   * NOTE: Requires real Razorpay credentials for live gateway refund processing.
   *       With mock credentials the DB status is updated but no real gateway call is made.
   */
  refundPayment: async (orderId, reason = '') => {
    const response = await api.post(`/payments/refund/${orderId}`, { reason });
    return response.data;
  },
};

export default paymentService;

File: backend/src/main/java/com/shopstack/service/PaymentService.java
package com.shopstack.service;

import com.shopstack.dto.payment.PaymentCreateRequest;
import com.shopstack.dto.payment.PaymentRefundRequest;
import com.shopstack.dto.payment.PaymentResponse;
import com.shopstack.dto.payment.PaymentVerifyRequest;

import java.util.List;

/**
 * PaymentService — Service interface for payment lifecycle operations.
 */
public interface PaymentService {

    /**
     * Creates a Razorpay order and stores a payment record for the given order.
     *
     * @param orderId  the ID of the order to pay
     * @param request  the payment creation request
     * @return the created {@link PaymentResponse}
     */
    PaymentResponse createPayment(Long orderId, PaymentCreateRequest request);

    /**
     * Verifies the Razorpay payment signature and updates the payment status.
     *
     * @param request  the verification request containing Razorpay IDs and signature
     * @return the updated {@link PaymentResponse}
     */
    PaymentResponse verifyPayment(PaymentVerifyRequest request);

    /**
     * Retrieves payment details for the specified order.
     *
     * @param orderId  the order ID
     * @return the {@link PaymentResponse} for the order
     */
    PaymentResponse getPaymentByOrder(Long orderId);

    /**
     * Refunds a successful payment for the specified order.
     *
     * @param orderId  the order ID
     * @param request  the refund request (optional reason)
     * @return the updated {@link PaymentResponse} with REFUNDED status
     */
    PaymentResponse refundPayment(Long orderId, PaymentRefundRequest request);

    /**
     * Returns all payment records (Admin-only).
     *
     * @return list of all {@link PaymentResponse}
     */
    List<PaymentResponse> getAllPayments();

    /**
     * Returns a single payment by its primary key (Admin-only).
     *
     * @param id  the payment ID
     * @return the {@link PaymentResponse}
     */
    PaymentResponse getPaymentById(Long id);
}
